lib/hashie/extensions/mash/underscore_keys.rb

```diff
@@ -0,0 +1,40 @@
+module Hashie
+  module Extensions
+    module Mash
+      # Overrides the indifferent access of a Mash to keep keys in
+      # underscore format.
+      #
+      # @example
+      #   class UnderscoreMash < ::Hashie::Mash
+      #     include Hashie::Extensions::Mash::UnderscoreKeys
+      #   end
+      #
+      #   mash = UnderscoreMash.new(symbolKey: { dataFrom: { java: true, javaScript: true } })
+      #   mash.symbol_key.data_from.java #=> true
+      #   mash.symbolKey.dataFrom.java_script  #=> true
+      module UnderscoreKeys
+        def self.included(base)
+          raise ArgumentError, "#{base} must descent from Hashie::Mash" unless base <= Hashie::Mash
+        end
+
+        private
+
+        def _underscore(string)
```

Comment:
This seems quite a bit different from, say, https://github.com/rails/rails/blob/04972d9b9ef60796dc8f0917817b5392d61fcf09/activesupport/lib/active_support/inflector/methods.rb#L96, what are the differences?

Reply:
I guess I went pretty old school and grabbed the version from 3.1 haha
https://github.com/rails/rails/blob/4dacedf983257aef38a8ebedb2d9a9c8fead8238/activesupport/lib/active_support/inflector/methods.rb#L48

I can update it to reflect the new version, I think there are likely performance and edge case improvements.

The one difference to this method that I added was converting spaces to underscores as well.

Reply:
Are you sure we want a different "underscore" behavior than ActiveSupport? Why?

Reply:
Hmm, I was thinking this could be slightly different than Rails. The intent here is making method names compatible with the Ruby's method naming conversions. `createdAt` -> `created_at`

But I think it could also go a step further, `created at` also does not follow the method naming convention, as a space is an illegal character. So my thought was we could convert them to underscore as well.

Maybe this should be called `include Hashie::Extensions::Mash::MethodSafeKeys`. What do you think?

Reply:
What do you think about implementing a generic extension that lets me transform keys whichever way I want (`TransformKeys`) as a custom behavior, and then deriving it into a bunch of specific behaviors such as a rails-compatible `UnderscoreKeys`?